core/crypto/kem/hybrid/hybrid.go

```diff
@@ -0,0 +1,341 @@
+	oth, ok := other.(*PublicKey)
+	if !ok {
+		return false
+	}
+	if pk.first == nil && pk.second == nil && oth.first == nil && oth.second == nil {
+		return true
+	}
+	if pk.first == nil || pk.second == nil || oth.first == nil || oth.second == nil {
+		return false
+	}
+	return pk.first.Equal(oth.first) && pk.second.Equal(oth.second)
+}
+
+func (pk *PublicKey) MarshalBinary() ([]byte, error) {
+	if pk.first == nil || pk.second == nil {
+		return nil, ErrUninitialized
+	}
+	first, err := pk.first.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	second, err := pk.second.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sch *Scheme) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
+	pk1, sk1, err := sch.first.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+	pk2, sk2, err := sch.second.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}, nil
+}
+
+func (sch *Scheme) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
+	if len(seed) != sch.SeedSize() {
+		panic(kem.ErrSeedSize)
+	}
+	h := sha3.NewShake256()
+	_, _ = h.Write(seed)
+	first := make([]byte, sch.first.SeedSize())
+	second := make([]byte, sch.second.SeedSize())
+	_, _ = h.Read(first)
+	_, _ = h.Read(second)
+
+	pk1, sk1 := sch.first.DeriveKeyPair(first)
+	pk2, sk2 := sch.second.DeriveKeyPair(second)
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}
+}
+
+func splitPRF(ss1, ss2, ct1, ct2 []byte) []byte {
+	// implement split PRF KEM combiner as:
+	// H(ss1 || ss2 || ct1 || ct2)
+	// in order to retain IND-CCA2 security
+	// as described here:
+	// https://link.springer.com/chapter/10.1007/978-3-319-76578-5_7
+
+	h, err := blake2b.New256(nil)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ss1)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ss2)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ct1)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ct2)
+	if err != nil {
+		panic(err)
+	}
+	return h.Sum(nil)
+}
+
+func (sch *Scheme) Encapsulate(pk kem.PublicKey) (ct, ss []byte, err error) {
+	pub, ok := pk.(*PublicKey)
+	if !ok {
+		return nil, nil, kem.ErrTypeMismatch
+	}
+
+	ct1, ss1, err := sch.first.Encapsulate(pub.first)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ct2, ss2, err := sch.second.Encapsulate(pub.second)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ss = splitPRF(ss1, ss2, ct1, ct2)
+
+	return append(ct1, ct2...), ss, nil
+}
+
+func (sch *Scheme) EncapsulateDeterministically(
+	pk kem.PublicKey, seed []byte,
+) (ct, ss []byte, err error) {
+	if len(seed) != sch.EncapsulationSeedSize() {
+		return nil, nil, kem.ErrSeedSize
+	}
+
+	h := sha3.NewShake256()
+	_, _ = h.Write(seed)
+	first := make([]byte, sch.first.EncapsulationSeedSize())
+	second := make([]byte, sch.second.EncapsulationSeedSize())
+	_, _ = h.Read(first)
+	_, _ = h.Read(second)
```

Comment:
You'll notice there's a LICENSE file in this directory. That's because I didn't write it... but I'm making a few small modifications to it so that the types are exported and so that it's IND-CCA2 instead of IND-CPA.

Please see cloudflare's original hybrid kem combiner, here: https://github.com/cloudflare/circl/blob/main/kem/hybrid/hybrid.go#L262-L290

Reply:
Okay. Fair enough. I'll fix this.

Reply:
fixed forever!

Reply:
`blake2b.XOF` is definitely better than `sha3` here, but I still think we could just ask for a larger random chunk (as described in other comment in this review), then the caller can deal with all the direct inputs.

Reply:
We now have `EncapsulationSeedSize() = sch.first.EncapsulationSeedSize() + sch.second.EncapsulationSeedSize()`
so we can delete the XOF and hash stuff and just call:
```go
	ct1, ss1, err := sch.first.EncapsulateDeterministically(pub.first, seed[ : sch.first.EncapsulationSeedSize() ])
// ...
	ct2, ss2, err := sch.second.EncapsulateDeterministically(pub.second, seed[ sch.first.EncapsulationSeedSize() : ])
```